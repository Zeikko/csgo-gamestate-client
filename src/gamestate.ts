import * as t from 'io-ts'

const GameStateProviderCodec = t.type({
  name: t.string,
  appid: t.number,
  version: t.number,
  steamid: t.string,
  timestamp: t.number,
})
export type GameStateProvider = t.TypeOf<typeof GameStateProviderCodec>

const GameStateTeamCodec = t.type({
  score: t.number,
  name: t.string,
  flag: t.string,
  consecutive_round_losses: t.number,
  timeouts_remaining: t.number,
  matches_won_this_series: t.number,
})
export type GameStateTeam = t.TypeOf<typeof GameStateTeamCodec>

const GameStateModeCodec = t.union([t.literal('casual'), t.literal('competitive')])
export type GameStateMode = t.TypeOf<typeof GameStateModeCodec>

const GameStateMapPhaseCodec = t.union([
  t.literal('live'),
  t.literal('intermission'),
  t.literal('gameover'),
  t.literal('warmup'),
])
export type GameStateMapPhase = t.TypeOf<typeof GameStateMapPhaseCodec>

const GameStateMapCodec = t.type({
  mode: GameStateModeCodec,
  name: t.string,
  phase: GameStateMapPhaseCodec,
  round: t.number,
  team_ct: GameStateTeamCodec,
  team_t: GameStateTeamCodec,
  num_matches_to_win_series: t.number,
  current_spectators: t.number,
  souvenirs_total: t.number,
})
export type GameStateMap = t.TypeOf<typeof GameStateMapCodec>

const GameStateRoundPhaseCodec = t.union([t.literal('live'), t.literal('freezetime'), t.literal('over')])
export type GameStateRoundPhase = t.TypeOf<typeof GameStateRoundPhaseCodec>

const GameStateRoundType = t.type({
  phase: GameStateRoundPhaseCodec,
})
const GameStateRoundStatePartial = t.partial({
  bomb: t.string,
})
const GameStateRoundCodec = t.intersection([GameStateRoundType, GameStateRoundStatePartial])
export type GameStateRound = t.TypeOf<typeof GameStateRoundCodec>

const GameStateTeamStringCodec = t.union([t.literal('T'), t.literal('CT')])
export type GameStateTeamString = t.TypeOf<typeof GameStateTeamStringCodec>

const GameStateActivityCodec = t.union([
  t.literal('playing'),
  t.literal('free'),
  t.literal('textinput'),
  t.literal('menu'),
])
export type GameStateActivity = t.TypeOf<typeof GameStateActivityCodec>

const GameStatePlayerStateType = t.type({
  health: t.number,
  armor: t.number,
  helmet: t.boolean,
  flashed: t.number,
  burning: t.number,
  smoked: t.number,
  money: t.number,
  round_kills: t.number,
  round_killhs: t.number,
  round_totaldmg: t.number,
  equip_value: t.number,
})
const GameStatePlayerStatePartial = t.partial({
  defusekit: t.boolean,
})
const GameStatePlayerStateCodec = t.intersection([GameStatePlayerStateType, GameStatePlayerStatePartial])
export type GameStatePlayerState = t.TypeOf<typeof GameStatePlayerStateCodec>

const GameStatePlayerCodec = t.type({
  steamid: t.string,
  name: t.string,
  activity: GameStateActivityCodec,
  state: GameStatePlayerStateCodec,
})
export type GameStatePlayer = t.TypeOf<typeof GameStatePlayerCodec>

const GameStateAllPlayerStateType = t.type({
  health: t.number,
  armor: t.number,
  helmet: t.boolean,
  flashed: t.number,
  burning: t.number,
  money: t.number,
  round_kills: t.number,
  round_killhs: t.number,
  round_totaldmg: t.number,
  equip_value: t.number,
})
const GameStateAllPlayerStatePartial = t.partial({
  defusekit: t.boolean,
})
const GameStateAllPlayerStateCodec = t.intersection([GameStateAllPlayerStateType, GameStateAllPlayerStatePartial])
export type GameStateAllPlayerState = t.TypeOf<typeof GameStateAllPlayerStateCodec>

const GameStateAllPlayerMatchStatsCodec = t.type({
  kills: t.number,
  assists: t.number,
  deaths: t.number,
  mvps: t.number,
  score: t.number,
})
export type GameStateAllPlayerMatchStats = t.TypeOf<typeof GameStateAllPlayerMatchStatsCodec>

const GameStateWeaponStateCodec = t.union([t.literal('active'), t.literal('holstered')])
export type GameStateWeaponState = t.TypeOf<typeof GameStateWeaponStateCodec>

const GameStateWeaponType = t.type({
  name: t.string,
  paintkit: t.string,
  type: t.string,
  state: GameStateWeaponStateCodec,
})
const GameStateWeaponPartial = t.partial({
  ammo_clip: t.number,
  ammo_clip_max: t.number,
  ammo_reserve: t.number,
})
const GameStateWeaponCodec = t.intersection([GameStateWeaponType, GameStateWeaponPartial])
export type GameStateWeapon = t.TypeOf<typeof GameStateWeaponCodec>

const GameStateAllPlayerWeaponsCodec = t.partial({
  weapon_0: GameStateWeaponCodec,
  weapon_1: GameStateWeaponCodec,
  weapon_2: GameStateWeaponCodec,
  weapon_3: GameStateWeaponCodec,
  weapon_4: GameStateWeaponCodec,
  weapon_5: GameStateWeaponCodec,
  weapon_6: GameStateWeaponCodec,
})
export type GameStateAllPlayerWeapons = t.TypeOf<typeof GameStateAllPlayerWeaponsCodec>

const GameStateAllPlayerCodec = t.type({
  name: t.string,
  observer_slot: t.number,
  team: GameStateTeamStringCodec,
  state: GameStateAllPlayerStateCodec,
  match_stats: GameStateAllPlayerMatchStatsCodec,
  weapons: GameStateAllPlayerWeaponsCodec,
  position: t.string,
  forward: t.string,
})
export type GameStateAllPlayer = t.TypeOf<typeof GameStateAllPlayerCodec>

const GameStatePhaseCodec = t.union([
  t.literal('freezetime'),
  t.literal('bomb'),
  t.literal('live'),
  t.literal('over'),
  t.literal('defuse'),
  t.literal('paused'),
  t.literal('timeout_ct'),
  t.literal('timeout_t'),
])
export type GameStatePhase = t.TypeOf<typeof GameStatePhaseCodec>

const GameStatePhaseCountdownCodec = t.type({
  phase: GameStatePhaseCodec,
  phase_ends_in: t.string,
})
export type GameStatePhaseCountdown = t.TypeOf<typeof GameStatePhaseCountdownCodec>

const GameStateGrenadeNameCodec = t.union([
  t.literal('weapon_smokegrenade'),
  t.literal('weapon_decoy'),
  t.literal('weapon_hegrenade'),
  t.literal('weapon_incgrenade'),
  t.literal('weapon_molotov'),
])
export type GameStateGrenadeName = t.TypeOf<typeof GameStateGrenadeNameCodec>

const GameStateGrenadeType = t.type({
  owner: t.number,
  lifetime: t.string,
  type: t.string,
})
const GameStateGrenadePartial = t.partial({
  effecttime: t.string,
  position: t.string,
  velocity: t.string,
})
const GameStateGrenadeCodec = t.intersection([GameStateGrenadeType, GameStateGrenadePartial])
export type GameStateGrenade = t.TypeOf<typeof GameStateGrenadeCodec>

export const GameStateCodec = t.type({
  provider: GameStateProviderCodec,
  map: GameStateMapCodec,
  round: GameStateRoundCodec,
  player: GameStatePlayerCodec,
  allplayers: t.record(t.string, GameStateAllPlayerCodec),
  phase_countdowns: GameStatePhaseCountdownCodec,
  grenades: t.record(t.string, GameStateGrenadeCodec),
})
export type GameState = t.TypeOf<typeof GameStateCodec>
