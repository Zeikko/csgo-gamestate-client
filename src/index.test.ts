import { parseGameState } from './index'
import { playingGameState } from './fixtures/gamestate'

describe('parseGameState', () => {
  it('returns parses gamestate', () => {
    expect(parseGameState(playingGameState)).toMatchSnapshot()
  })

  it('fails', () => {
    expect(() => parseGameState('{ "bad": "input" }')).toThrow(
      'Expecting { name: string, appid: number, version: number, steamid: string, timestamp: number } at provider but instead got: undefined',
    )
  })
})
