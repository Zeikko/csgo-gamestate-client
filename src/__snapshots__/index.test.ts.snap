// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseGameState returns parses gamestate 1`] = `
Object {
  "allplayers": Object {
    "76561197961610974": Object {
      "forward": "-1.00, 0.01, 0.07",
      "match_stats": Object {
        "assists": 3,
        "deaths": 19,
        "kills": 22,
        "mvps": 3,
        "score": 52,
      },
      "name": "allu",
      "observer_slot": 1,
      "position": "276.59, -1576.46, -175.97",
      "state": Object {
        "armor": 100,
        "burning": 0,
        "equip_value": 6850,
        "flashed": 0,
        "health": 100,
        "helmet": true,
        "money": 500,
        "round_killhs": 0,
        "round_kills": 3,
        "round_totaldmg": 222,
      },
      "team": "T",
      "weapons": Object {
        "weapon_0": Object {
          "name": "weapon_bayonet",
          "paintkit": "aq_steel_knife",
          "state": "holstered",
          "type": "Knife",
        },
        "weapon_1": Object {
          "ammo_clip": 20,
          "ammo_clip_max": 20,
          "ammo_reserve": 120,
          "name": "weapon_glock",
          "paintkit": "aa_glock_18_urban_moon_fever",
          "state": "holstered",
          "type": "Pistol",
        },
        "weapon_2": Object {
          "ammo_clip": 7,
          "ammo_clip_max": 10,
          "ammo_reserve": 30,
          "name": "weapon_awp",
          "paintkit": "hy_ddpat_pink",
          "state": "active",
          "type": "SniperRifle",
        },
        "weapon_3": Object {
          "ammo_reserve": 1,
          "name": "weapon_molotov",
          "paintkit": "default",
          "state": "holstered",
          "type": "Grenade",
        },
      },
    },
    "76561197962710908": Object {
      "forward": "-1.00, 0.01, 0.06",
      "match_stats": Object {
        "assists": 0,
        "deaths": 19,
        "kills": 18,
        "mvps": 3,
        "score": 45,
      },
      "name": "Aerial",
      "observer_slot": 4,
      "position": "-149.04, -2181.76, -149.20",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 350,
        "round_killhs": 0,
        "round_kills": 0,
        "round_totaldmg": 0,
      },
      "team": "T",
      "weapons": Object {},
    },
    "76561197993474578": Object {
      "forward": "0.16, -0.96, -0.24",
      "match_stats": Object {
        "assists": 5,
        "deaths": 19,
        "kills": 20,
        "mvps": 3,
        "score": 56,
      },
      "name": "fitch",
      "observer_slot": 9,
      "position": "-1341.48, -2374.57, -210.49",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 6150,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 2800,
        "round_killhs": 0,
        "round_kills": 0,
        "round_totaldmg": 0,
      },
      "team": "CT",
      "weapons": Object {},
    },
    "76561198013172503": Object {
      "forward": "-1.00, 0.01, 0.06",
      "match_stats": Object {
        "assists": 6,
        "deaths": 17,
        "kills": 21,
        "mvps": 7,
        "score": 71,
      },
      "name": "xseveN",
      "observer_slot": 5,
      "position": "245.37, -1493.42, -175.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 1250,
        "round_killhs": 1,
        "round_kills": 1,
        "round_totaldmg": 178,
      },
      "team": "T",
      "weapons": Object {},
    },
    "76561198013243326": Object {
      "forward": "-1.00, 0.01, 0.06",
      "match_stats": Object {
        "assists": 3,
        "deaths": 21,
        "kills": 17,
        "mvps": 1,
        "score": 46,
      },
      "name": "Aleksib",
      "observer_slot": 3,
      "position": "-807.99, -1056.00, -167.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 500,
        "round_killhs": 0,
        "round_kills": 0,
        "round_totaldmg": 0,
      },
      "team": "T",
      "weapons": Object {},
    },
    "76561198027839825": Object {
      "forward": "-1.00, 0.01, 0.06",
      "match_stats": Object {
        "assists": 6,
        "deaths": 23,
        "kills": 16,
        "mvps": 1,
        "score": 43,
      },
      "name": "sergej",
      "observer_slot": 2,
      "position": "68.76, -2107.69, -35.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 1550,
        "round_killhs": 0,
        "round_kills": 0,
        "round_totaldmg": 0,
      },
      "team": "T",
      "weapons": Object {},
    },
    "76561198036125584": Object {
      "forward": "0.16, -0.96, -0.24",
      "match_stats": Object {
        "assists": 2,
        "deaths": 17,
        "kills": 21,
        "mvps": 3,
        "score": 55,
      },
      "name": "Jame",
      "observer_slot": 6,
      "position": "-201.84, -2206.60, -167.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 7950,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 2150,
        "round_killhs": 0,
        "round_kills": 1,
        "round_totaldmg": 100,
      },
      "team": "CT",
      "weapons": Object {},
    },
    "76561198067937899": Object {
      "forward": "0.15, -0.96, -0.24",
      "match_stats": Object {
        "assists": 5,
        "deaths": 20,
        "kills": 16,
        "mvps": 3,
        "score": 44,
      },
      "name": "KrizzeN",
      "observer_slot": 8,
      "position": "-463.08, -2026.80, -179.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 6150,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 3400,
        "round_killhs": 1,
        "round_kills": 2,
        "round_totaldmg": 200,
      },
      "team": "CT",
      "weapons": Object {},
    },
    "76561198127236290": Object {
      "forward": "0.16, -0.96, -0.24",
      "match_stats": Object {
        "assists": 1,
        "deaths": 19,
        "kills": 23,
        "mvps": 0,
        "score": 53,
      },
      "name": "qikert",
      "observer_slot": 0,
      "position": "-1296.48, -1174.83, -167.97",
      "state": Object {
        "armor": 0,
        "burning": 0,
        "equip_value": 5300,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 7250,
        "round_killhs": 0,
        "round_kills": 0,
        "round_totaldmg": 0,
      },
      "team": "CT",
      "weapons": Object {},
    },
    "76561198173201923": Object {
      "forward": "0.15, -0.96, -0.24",
      "match_stats": Object {
        "assists": 5,
        "deaths": 19,
        "kills": 19,
        "mvps": 3,
        "score": 52,
      },
      "name": "buster",
      "observer_slot": 7,
      "position": "-438.32, -1797.39, -174.37",
      "state": Object {
        "armor": 100,
        "burning": 0,
        "defusekit": true,
        "equip_value": 7750,
        "flashed": 0,
        "health": 100,
        "helmet": true,
        "money": 6250,
        "round_killhs": 0,
        "round_kills": 1,
        "round_totaldmg": 100,
      },
      "team": "CT",
      "weapons": Object {
        "weapon_0": Object {
          "name": "weapon_knife_m9_bayonet",
          "paintkit": "am_zebra",
          "state": "active",
          "type": "Knife",
        },
        "weapon_1": Object {
          "ammo_clip": 12,
          "ammo_clip_max": 12,
          "ammo_reserve": 24,
          "name": "weapon_usp_silencer",
          "paintkit": "cu_usp_elegant",
          "state": "holstered",
          "type": "Pistol",
        },
        "weapon_2": Object {
          "ammo_clip": 5,
          "ammo_clip_max": 10,
          "ammo_reserve": 30,
          "name": "weapon_awp",
          "paintkit": "cu_awp_asimov",
          "state": "holstered",
          "type": "SniperRifle",
        },
        "weapon_3": Object {
          "ammo_reserve": 1,
          "name": "weapon_hegrenade",
          "paintkit": "default",
          "state": "holstered",
          "type": "Grenade",
        },
      },
    },
  },
  "grenades": Object {},
  "map": Object {
    "current_spectators": 298079,
    "mode": "competitive",
    "name": "workshop/1729357842/de_mirage",
    "num_matches_to_win_series": 2,
    "phase": "live",
    "round": 27,
    "souvenirs_total": 0,
    "team_ct": Object {
      "consecutive_round_losses": 0,
      "flag": "KZ",
      "matches_won_this_series": 0,
      "name": "Avangar",
      "score": 12,
      "timeouts_remaining": 3,
    },
    "team_t": Object {
      "consecutive_round_losses": 0,
      "flag": "FI",
      "matches_won_this_series": 0,
      "name": "ENCE",
      "score": 15,
      "timeouts_remaining": 1,
    },
  },
  "phase_countdowns": Object {
    "phase": "bomb",
    "phase_ends_in": "2.6",
  },
  "player": Object {
    "activity": "playing",
    "name": "allu",
    "observer_slot": 1,
    "state": Object {
      "armor": 100,
      "burning": 0,
      "equip_value": 6850,
      "flashed": 0,
      "health": 100,
      "helmet": true,
      "money": 500,
      "round_killhs": 0,
      "round_kills": 3,
      "round_totaldmg": 222,
      "smoked": 0,
    },
    "steamid": "76561197961610974",
    "team": "T",
  },
  "previously": Object {
    "allplayers": Object {
      "76561197961610974": Object {
        "forward": "-0.84, -0.54, -0.04",
        "position": "295.79, -1590.00, -175.97",
        "weapons": Object {
          "weapon_0": Object {
            "state": "active",
          },
          "weapon_2": Object {
            "state": "holstered",
          },
        },
      },
      "76561197962710908": Object {
        "forward": "-0.84, -0.55, -0.04",
      },
      "76561197993474578": Object {
        "forward": "0.97, 0.06, -0.22",
      },
      "76561198013172503": Object {
        "forward": "-0.84, -0.55, -0.04",
      },
      "76561198013243326": Object {
        "forward": "-0.84, -0.54, -0.03",
      },
      "76561198027839825": Object {
        "forward": "-0.84, -0.55, -0.03",
      },
      "76561198036125584": Object {
        "forward": "0.97, 0.06, -0.22",
      },
      "76561198067937899": Object {
        "forward": "-0.80, 0.33, -0.51",
      },
      "76561198127236290": Object {
        "forward": "0.97, 0.06, -0.22",
      },
      "76561198173201923": Object {
        "forward": "0.97, 0.06, -0.22",
        "position": "-410.31, -1633.96, -39.97",
        "weapons": Object {
          "weapon_0": Object {
            "state": "holstered",
          },
          "weapon_2": Object {
            "ammo_clip": 6,
            "state": "active",
          },
        },
      },
    },
    "phase_countdowns": Object {
      "phase_ends_in": "3.8",
    },
  },
  "provider": Object {
    "appid": 730,
    "name": "Counter-Strike: Global Offensive",
    "steamid": "76561197962883058",
    "timestamp": 1594191099,
    "version": 13759,
  },
  "round": Object {
    "bomb": "planted",
    "phase": "live",
  },
}
`;
